### PR TITLE
Stabilize extended_key_value_attributes

### LINKED ISSUE #44732: Tracking issue: rfc 1990 - add external doc attribute to rustc
RFC PR: https://github.com/rust-lang/rfcs/pull/1990
RFC text: https://github.com/rust-lang/rfcs/blob/master/text/1990-external-doc-attribute.md

[Current documentation](https://doc.rust-lang.org/nightly/unstable-book/language-features/external-doc.html)

-----

Summary of how to use this:

1. Add `#![feature(external_doc)]` to your crate.
2. Add a file `"src/some-docs.md"` with some docs to your crate.
3. Add an attribute `#[doc(include = "some-docs.md")]` to something that needs some docs. The file path is relative to `lib.rs`, so if you want a `doc` folder to live alongside `src`, then all your paths inside the `doc(include)` attributes need to begin with `../doc`.

Summary of current status:

- Initial implementation landed in https://github.com/rust-lang/rust/pull/44781, on 2017-11-22.
- Updates based on conversation in this thread landed in https://github.com/rust-lang/rust/pull/46858, on 2017-12-21.
- Please use this in your own projects and comment in this thread if something goes wrong! (Or even if it goes well! Knowing it's working as intended is great!)

-----

Current tasks:

- [x] Land https://github.com/rust-lang/rust/pull/44781
- [x] ~~(promote the "failed to load file" warnings to proper lints?)~~ (Make them hard errors instead, per the RFC https://github.com/rust-lang/rust/pull/46858)
- [ ] Ensure line number info is properly preserved in doctest errors

### LINKED ISSUE #78835: Tracking Issue for `#![feature(extended_key_value_attributes)]`
Implemented in https://github.com/rust-lang/rust/pull/78837.

With this feature arbitrary expressions are supported in values of key-value attributes at parse time.
```
#[my_attr = EXPR]
```
Previously only unsuffixed literals and interpolated expressions (`$expr`) were accepted.

If the attribute in question survives expansion, then the value is still restricted to unsuffixed literals by a semantic check. This restriction may also be relaxed in the future.

### LINKED ISSUE #82768: `extended_key_value_attributes` cannot be enabled conditionally
I tried this code:

```rust
#![cfg_attr(
    any(),
    feature(extended_key_value_attributes),
    doc = include_str!("../README.md")
)]
```

I expected to see this happen: The compiler gives no error, because `doc = include_str!` should only be compiled if the predicate is true.

Instead, this happened: The compiler gives an error that `extended_key_value_attributes` is unstable.

I guess this happens because the compiler has to parse the attribute even if the cfg is false? @petrochenkov is giving an error here intentional?

### Meta

`rustc --version`: 1.52.0-nightly (2021-03-01 4f20caa6258d4c74ce6b)

cc https://github.com/rust-lang/rust/issues/78835
_Originally posted by @daxpedda in https://github.com/rust-lang/rust/issues/82539#issuecomment-790418123_

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/rust/issues/44732. Closes https://github.com/rust-lang/rust/issues/78835. Closes https://github.com/rust-lang/rust/issues/82768 (by making it irrelevant).

 # Stabilization report

 ## Summary

This stabilizes using macro expansion in key-value attributes, like so:

 ```rust
 #[doc = include_str!("my_doc.md")]
 struct S;
 ```

Eager macro expansion only happens in a limited number of built-in attributes, with `#[doc]` being the primary example that will benefit from this. Notably, the attributes `path`, `crate_type`, and `recursion_limit` do not support this because they need their values before or during expansion.  It is also not supported with proc-macros.

See Petrochenkov's excellent blog post [on internals](https://internals.rust-lang.org/t/macro-expansion-points-in-attributes/11455)
for alternatives that were considered and rejected ("why accept no more and no less?")

This has been available on nightly since 1.50 with no major issues.

### Accepted syntax

The parser accepts arbitrary Rust expressions in this position, but any expression other than a macro invocation will ultimately lead to an error because it is not expected by the built-in expression forms (e.g., `#[doc]`).  Note that decorators and the like may be able to observe other expression forms.

### Expansion ordering

Expansion of macro expressions in "inert" attributes occurs after decorators have executed, analogously to macro expressions appearing in the function body or other parts of decorator input.

There is currently no way for decorators to accept macros in key-value position if macro expansion must be performed before the decorator executes (if the macro can simply be copied into the output for later expansion, that can work).

## Test cases

 - https://github.com/rust-lang/rust/blob/master/src/test/ui/attributes/key-value-expansion-on-mac.rs
 - https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/external-doc.rs

The feature has also been dogfooded extensively in the compiler and
standard library:

- https://github.com/rust-lang/rust/pull/83329
- https://github.com/rust-lang/rust/pull/83230
- https://github.com/rust-lang/rust/pull/82641
- https://github.com/rust-lang/rust/pull/80534

## Implementation history

- Initial proposal: https://github.com/rust-lang/rust/issues/55414#issuecomment-554005412
- Experiment to see how much code it would break: https://github.com/rust-lang/rust/pull/67121
- Preliminary work to restrict expansion that would conflict with this
feature: https://github.com/rust-lang/rust/pull/77271
- Initial implementation: https://github.com/rust-lang/rust/pull/78837
- Fix for an ICE: https://github.com/rust-lang/rust/pull/80563

## Unresolved Questions

~~https://github.com/rust-lang/rust/pull/83366#issuecomment-805180738 listed some concerns, but they have been resolved as of this final report.~~

<a id="additional-information">

 ## Additional Information

</a>

 There are two workarounds that have a similar effect for `#[doc]`
attributes on nightly. One is to emulate this behavior by using a limited version of this feature that was stabilized for historical reasons:

```rust
macro_rules! forward_inner_docs {
    ($e:expr => $i:item) => {
        #[doc = $e]
        $i
    };
}

forward_inner_docs!(include_str!("lib.rs") => struct S {});
```

The other is to use `doc(include)`:

```rust
 #![feature(external_doc)]
 #[doc(include = "lib.rs")]
 struct S {}
```

The first works, but is non-trivial for people to discover, and
difficult to read and maintain. The second is a strange special-case for
a particular use of the macro. This generalizes it to work for any use
case, not just including files.

I plan to remove `doc(include)` when this is stabilized
(https://github.com/rust-lang/rust/pull/82539). The `forward_inner_docs`
workaround will still compile without warnings, but I expect it to be
used less once it's no longer necessary.

### CODE REVIEW

**Comment by jyn514** [comment]
r? @petrochenkov (forgot to put it in the PR description, sorry).

I added T-lang and T-compiler since I think they may need to both sign off on the FCP? Feel free to remove one or the other if necessary.

**Comment by pksunkara** [comment]
Stabilizing this would not make https://github.com/rust-lang/rust/issues/82768 irrelevant because of people trying to provide backward compatibility.

Clap's MSRV is 1.46.0. We would like to keep it the same and use the `doc` feature flag to conditionally enable the `doc = include_str!` when building docs. That way, MSRV is bumped to 1.53.0 only when building the docs.

**Comment by jyn514** [comment]
> Clap's MSRV is 1.46.0. We would like to keep it the same and use the doc feature flag to conditionally enable the doc = include_str! when building docs. That way, MSRV is bumped to 1.53.0 only when building the docs.

@pksunkara see the workaround in https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643.

**Comment by camelid** [comment]
> Closes #44732.

I don't think this should close that issue since this PR does not stabilize, deprecate, or remove `doc(include)`. #82539 should be the one to close that issue.

**Comment by petrochenkov** [comment]
>One is to emulate this behavior by using eager
expansion

This is not eager expansion, this is a limited version of the same `extended_key_value_attributes` that exists for historical reasons.

**Comment by petrochenkov** [comment]
There are two questions to answer/acknowledge before stabilizing this:

- Whether to stabilize `#[attr = ARBITRARY_EXPRESSION]` or macro calls only `#[attr = mac::call!(a b c)]`.

    If arbitrary expressions are stabilized, then things like
    ```rust
    #[attr = {
        let a = 10;
        println!("{}", a + 11);
    }]
    ```
    become legal as well, but only syntactically (under `cfg(FALSE)`), they will still be rejected semantically.

    Even if we only stabilize macros, right now the compiler has to know that `mac::call!(a b c)` is an *expression* to expand it, but it's not something fundamental, we may be able to expand macros in context independent way in the future.

- Relative order of expanding macros in `#[attr = mac!()]` and in other places.

    Right now `mac!()` is expanded last of all the macro invocations below
    ```rust
    #[macro_attr1]
    #[doc = mac!()]
    #[macro_attr2]
    #[derive(MacroDerive1, MacroDerive2)]
    struct S;

    bar!();
    ```
    and not between `macro_attr1` and `macro_attr2` as some people could expect.

    However, this falls under the general rule telling that the expansion order of macros *applied to different targets* is unspecified, i.e. `macro_attr1` and `macro_attr2` are applied to the same target `struct S` so they are ordered, but `mac!()` is an entirely separate invocation similarly to e.g. `bar!()`.
The problem is only that the expansion order is easily discoverable by `macro_attr(1,2)` and `MacroDerive(1,2)` here, so it becomes fixed in practice.

    Expanding `mac!()` in a way making its result visible to `macro_attr2` should be somewhat annoying from the implementation point of view (I didn't investigate details).

**Comment by camelid** [comment]
>   * Expanding `mac!()` in a way making its result visible to `macro_attr2` should be somewhat annoying from the implementation point of view (I didn't investigate details).

Do you mean `macro_attr1` or do attribute macros evaluate from the outermost to the innermost? I would have assumed `macro_attr1` would be applied on the result of `doc`; `doc` on the result of `macro_attr`; and `macro_attr2` on the result of `derive`.

**Comment by petrochenkov** [comment]
@camelid 
Macros on the same target are expanded in the left-to-right order, so `#[macro_attr1]` is run first and cannot see results of anything after it.

**Comment by camelid** [comment]
> @camelid
> Macros on the same target are expanded in the left-to-right order, so `#[macro_attr1]` is run first and cannot see results of anything after it.

Huh, I didn't realize that. I think I'll have to read the rustc-dev-guide chapter on expansion :)

**Comment by jyn514** [comment]
> I don't think this should close that issue since this PR does not stabilize, deprecate, or remove doc(include). #82539 should be the one to close that issue.

Stabilizing `extended_key_value_attributes` will stabilize including external files in documentation, which I took as the point of that issue. I don't think whether it uses `doc(include)` or `doc = include_str!` is particularly important - if I close it when #82539 merges then people still won't be able to use external documentation, which seems unsatisfactory. Similarly, closing it when `doc(include)` is made a hard error also seems unhelpful.

**Comment by kangalioo** [comment]
> Whether to stabilize `#[attr = ARBITRARY_EXPRESSION]` or macro calls only `#[attr = mac::call!(a b c)]`.

It would be amazing to have arbitrary expression support like that. Procedural attribute macros could implement more ergonomic and natural APIs without having to drop down to manual TokenStream parsing:
```rust
#[my_attr(numbers = [1, 2, 3])]
```
instead of
```rust
#[my_attr(numbers("1", "2", "3"))]
```

**Comment by jyn514** [comment]
> Whether to stabilize #[attr = ARBITRARY_EXPRESSION] or macro calls only #[attr = mac::call!(a b c)].

Arbitrary expressions can be added later and still be backwards compatible, right? I think we should do that to avoid delaying this longer than necessary.

**Comment by petrochenkov** [comment]
>Procedural attribute macros could implement more ergonomic and natural APIs without having to drop down to manual TokenStream parsing

Yes, dtolnay, for example, requested supporting paths in this position for use in cxx (https://github.com/rust-lang/rust/pull/76734).

**Comment by pksunkara** [comment]
Just to be clear, there are no blockers for stabilizing `#[attr = ARBITRARY_EXPRESSION]` as part of this, right?

**Comment by petrochenkov** [comment]
I'm not aware of any such blockers.

**Comment by workingjubilee** [comment]
@jyn514
Do you believe there has been adequate testing to support stabilizing expanding arbitrary expressions? Or do you believe it is insufficient enough to delay such? I would surmise that if it is highly requested, part of the accepted RFC, and reasonably well-tested, then it is probably time to stabilize.

**Comment by joshtriplett** [comment]
Proposing stabilization:

@rfcbot merge

**Comment by jyn514** [comment]
Oh ugh, I thought I included them in this PR, but the reference is in a submodule so it didn't get picked up by `git add`. I don't have access to the computer I originally wrote this on, I'll try to write up the changes some time in the next few weeks.

It's not much different from the example I gave, this allows `#[doc = MacroInvocation]` as long as `MacroInvocation` expands to a constant expression.

**Comment by nikomatsakis** [comment]
OK. An important point here is that this was already permitted in the specific case where an expr was interpolated in via macro argument, right? So we're actually rationalizing existing behavior?

One question I had for @petrochenkov, related to the expansion order:

> However, this falls under the general rule telling that the expansion order of macros applied to different targets is unspecified, i.e. macro_attr1 and macro_attr2 are applied to the same target struct S so they are ordered, but mac!() is an entirely separate invocation similarly to e.g. bar!().
> The problem is only that the expansion order is easily discoverable by macro_attr(1,2) and MacroDerive(1,2) here, so it becomes fixed in practice.

I wanted to say that these two statements are kind of in tension, right? i.e., we say that the order is undefined, but then that it can readily be observed.

Still, I'm trying to understand the constraints. I think (a) we cannot expand `mac!()` before `macro_attr1`, because `macro_attr1` will take in all the following tokens and may generate something completely different; (b) the `mac!()` macro appears in something that is "outside" of `macro_attr2`. How is it completely observed? I'm revealing my ignorance here of the details of which tokens get provided as inputs to decorators, I suppose. Do decorators get all the existing attributes as inputs?

**Comment by jyn514** [comment]
> An important point here is that this was already permitted in the specific case where an expr was interpolated in via macro argument, right? So we're actually rationalizing existing behavior?

Yes. You can get exactly the behavior in this PR on stable by introducing a `forward_inner_attrs` macro and using that everywhere.

**Comment by nikomatsakis** [comment]
> Perhaps I wasn't right about the order being undefined and the mac!() case is more or less the same as

I believe this is a correct analogy, @petrochenkov

**Comment by nikomatsakis** [comment]
@petrochenkov can you extend the stabilization report to cover the expected ordering and the analogy to fn bodies?

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
@rfcbot concern ordering

I'd like to see it documented that macro invocations in "inert" attributes are not expanded until after procedural macros have executed, analogously to macro invocations appearing in the function body.

**Comment by pnkfelix** [comment]
@jyn514 and @petrochenkov : the lang team spent a while discussing this PR in our triage meeting today.

We were not clear on a few details about what is intended to be stabilized here.

Specifically, we were not clear about what the current answers are to the [questions posed by @petrochenkov](https://github.com/rust-lang/rust/pull/83366#issuecomment-805180738).

I've included some discussion of what I did to try to answer the questions myself below.

But, the point of my feedback isn't the specifics of what I did just now. The main point is that we think the stabilization report needs to be extended with answers to the questions.


----

As an example of the kind of question that still needs a clear answer in the stabilization report: From brief skimming of the source code during the meeting, it *seems* like support for parsing of arbitrary expressions is being stabilized (which, if true, might give us pause on accepting stabilization).

I tried to use the dialogue here on the PR to try to infer what the expected situation is, but even then it is not clear, at least to me. [This comment from jyn514](https://github.com/rust-lang/rust/pull/83366#issuecomment-812925757) implies to me that the code does not yet support arbitrary expressions. But then [this later comment from workingjubilee](https://github.com/rust-lang/rust/pull/83366#issuecomment-815434464) implies that arbitrary expressions **are** supported?

I finally resorted to exploring the behavior in the [playpen](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0616968690a5ad9b57f0ce7b67ba5ed1).

From my experiments in the playpen, it seems like the feature flag is behaving in exactly the way that @petrochenkov described under the "arbitrary expressions are stabilized": namely, arbitrary expressions are *parsed* in the expr position of `#[key = expr]`.

(I think some members of the lang team were expecting the parsing behavior here to be more restricted than that, so that parser in this context would only accept a subset of the rust grammar that would be something like `SE ::= string-lit | macro!(SE, ...)`. But that's largely a guess on my part.)

Obviously most arbitrary expressions won't expand into a string literal, and thus will be semantically rejected after parsing is finished. (I think the concern is that we'd be stepping into having to support *parsing* of arbitrary expressions, at least ones that end up being ignored.)

**Comment by petrochenkov** [comment]
>Specifically, we were not clear about what the current answers are to the questions posed by @petrochenkov.

>Whether to stabilize `#[attr = ARBITRARY_EXPRESSION]` or macro calls only `#[attr = mac::call!(a b c)]`.

This PR stabilizes `#[attr = ARBITRARY_EXPRESSION]`.

>Relative order of expanding macros in `#[attr = mac!()]` and in other places.

With this PR `mac!()` is expanded after `#[attr]`'s macro attribute neighbors.

---

>I think the concern is that we'd be stepping into having to support *parsing* of arbitrary expressions, at least ones that end up being ignored.

They are not necessarily ignored, they can be processed and removed by proc macros, like suggested in #76734, for example.

**Comment by nikomatsakis** [comment]
Personally I feel comfortable supporting arbitrary expressions. Are there particular complications that we should be aware of? One question that was raised was whether there might be interactions with hygiene -- e.g., if identifiers were inserted into those expressions.

**Comment by joshtriplett** [comment]
I'm comfortable supporting arbitrary expressions as well.

**Comment by scottmcm** [comment]
One thing I wanted to clarify (and sorry if I missed it specified elsewhere) about

> Whether to stabilize `#[attr = ARBITRARY_EXPRESSION]`

Is this parsing an `expr`ession or a `tt*` in that position?  For example, would it allow `#[attr = +(3)]`?

**Comment by petrochenkov** [comment]
@scottmcm 
>Is this parsing an `expr`ession or a `tt*` in that position?  For example, would it allow `#[attr = +(3)]`?

An expression.
The example won't be allowed because we don't have a unary `+` operator.

**Comment by nikomatsakis** [comment]
To be clear, though, this error occurs at the point where we *parse* the expression, right?

It would *lex* and potentially be seen by a procedural macro, wouldn't it?

**Comment by petrochenkov** [comment]
>To be clear, though, this error occurs at the point where we parse the expression, right?

Correct.

>It would lex

Even less doubt in this.

>and potentially be seen by a procedural macro

No (or yes but only as a part of error recovery).
Basic attribute syntax is parsed before any expansion, so in the `+(3)` example we get a parsing error before expanding anything.
Input to attribute macros (as opposed to fn-like macros) must already be parse-able Rust code.

**Comment by nikomatsakis** [comment]
@petrochenkov 

>  Input to attribute macros (as opposed to fn-like macros) must already be parse-able Rust code.

I see

**Comment by petrochenkov** [comment]
@nikomatsakis 
Should the "ordering" concern be resolved now?

**Comment by nikomatsakis** [comment]
@petrochenkov I wanted to see those changes moved to the main PR description for posterity, but I suppose I can make those edits myself. =) I'm not sure if others have concerns.

**Comment by nikomatsakis** [comment]
@rfcbot resolve 

ok, I edited the OP to add two clarifications and notes. I'm going to resolve my concern. I do recall that @cramertj was concerned about the difference that this change creates between builtin attributes and what decorators can do.

In particular, if you have a derive (or whatever) with a custom attribute like `#[foo = "bar"]`, you can't do:

```rust
#[derive(MyDerive)]
struct Foo {
    #[foo = include_str!(...)]
    x: u32
}
```

unless you are able to pipe the `include_str!` through to your output. 

I *think* the conclusion at the meeting was "oh, well, just one more use case for early expansion", though.

**Comment by nikomatsakis** [comment]
@rfcbot resolve ordering

**Comment by petrochenkov** [comment]
@jyn514 
(Entering FCP, not finishing FCP, I've updated the labels.)

**Comment by nagisa** [comment]
Hm, it seems like `rustfmt` is not taking kindly to this feature. E.g.

```
error: unexpected token: `include_str`
 --> src/lib.rs:5:10
  |
5 | #![doc = include_str!("../README.md")]
  |          ^^^^^^^^^^^
```

is output from `cargo fmt`. Not sure if we want to block fcp on it or not.

**Comment by camelid** [comment]
Also, it may be worth noting that the errors are a bit off:

```
error: couldn't read src/../README.md: No such file or directory (os error 2)
 --> src/lib.rs:2:10
  |
2 | #![doc = include_str!("../README.md")]
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected token: `(/*ERROR*/)`
 --> src/lib.rs:2:10
  |
2 | #![doc = include_str!("../README.md")]
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors
```

I don't think it's worth blocking on, but maybe it could instead use a `delay_span_bug` when it encounters an error token?

**Comment by camelid** [comment]
> Hm, it seems like `rustfmt` is not taking kindly to this feature. E.g.
> 
> ```
> error: unexpected token: `include_str`
>  --> src/lib.rs:5:10
>   |
> 5 | #![doc = include_str!("../README.md")]
>   |          ^^^^^^^^^^^
> ```
> 
> is output from `cargo fmt`. Not sure if we want to block fcp on it or not.

(Note for people following along: the associated issue \[#4834\] was closed due to the rustfmt version being old.)

**Comment by jyn514** [comment]
@camelid can you open a separate issue for that?

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -281,6 +281,8 @@ declare_features! (
     (accepted, or_patterns, "1.53.0", Some(54883), None),
     /// Allows defining identifiers beyond ASCII.
     (accepted, non_ascii_idents, "1.53.0", Some(55467), None),
+    /// Allows arbitrary expressions in key-value attributes at parse time.
+    (accepted, extended_key_value_attributes, "1.53.0", Some(78835), None),
```
Comment (petrochenkov) [inline compiler/rustc_feature/src/accepted.rs:285]:
```suggestion
    (accepted, extended_key_value_attributes, "1.54.0", Some(78835), None),
```

**Comment by petrochenkov** [comment]
r=me after rebasing and updating the stabilization version.

**Comment by nikomatsakis** [comment]
So I was thinking -- I think we should be tracking any such stabilization and making a point to extend the reference afterwards. @jyn514 are you ok taking that work on? :) I'm not sure how much existing material is there. In any case, I'm going to add this to a "needs reference" column in our lang team project board.

**Comment by jyn514** [comment]
@nikomatsakis the next few weeks my schedule is very erratic; I can try but I can't promise when I'll have time.

**Comment by jyn514** [comment]
@nikomatsakis not sure how milestones work but if you could assign me to an issue tracking the update to the reference, that would be very helpful, I'll definitely forget otherwise 😅

**Comment by nikomatsakis** [comment]
@jyn514 ok, I opened rust-lang/reference#1028, thanks!

**Comment by webern** [comment]
I came here from the [release notes](https://blog.rust-lang.org/2021/07/29/Rust-1.54.0.html) for 1.54. Does this feature work for derive macros? It doesn't seem like it does. I'm trying something that looks like this:

```rust
#[some_derive_macro(some_attribute = myfoo!())]
struct Whatever {
    
}
```

Which does not work with `Unable to parse attribute: expected literal`.

What I *actually* want is this: https://github.com/rust-lang/rust/issues/52393

And would look like this:

```rust

const MY_FOO: &str = "bloop";

#[some_derive_macro(some_attribute = MY_FOO)]
struct Whatever {
    
}
```

I thought I might be able to do a workaround using the new 1.54 feature:

```rust
macro_rules identity! { ($x:expr) => { $x }; }

const MY_FOO: &str = "bloop";

#[some_derive_macro(some_attribute = identity!(MY_FOO))]
struct Whatever {
    
}
```

Is there any chance that derive macros can get the same love as proc macros so that we can use constants as values in derive macro attribute values?

**Comment by jplatte** [comment]
This is not specifically about proc-macros: https://github.com/rust-lang/rust/issues/78835#issuecomment-845831023

**Comment by webern** [comment]
> This is not specifically about proc-macros: [#78835 (comment)](https://github.com/rust-lang/rust/issues/78835#issuecomment-845831023)

So if I read that correctly, this is supported:

```rust
#[some_derive_macro = something!()]
struct X {}
```

Bot not this:

```rust
#[some_derive_macro(attribute1 = something!())]
struct X {}
```

Is that what https://github.com/rust-lang/rust/issues/78835#issuecomment-845831023 is saying?

Thanks.

**Comment by jyn514** [comment]
@webern Yes.

**Comment by michalfita** [comment]
@jyn514 but that doesn't:
```rust
#[path = concat!(env!("OUT_DIR"), "/lib.rs")]
```

**Pull request merged**